core/src/main/java/discord4j/core/GatewayDiscordClient.java

```diff
@@ -562,7 +563,15 @@ private GuildData toGuildData(GuildUpdateData guild) {
                                 .from(request)
                                 .nonce(nonce)
                                 .build(), shardId)))
-                .thenMany(Flux.defer(incomingMembers));
+                .thenMany(Flux.defer(incomingMembers))
+                .doOnComplete(() -> {
+                    if (request.query().toOptional().map(String::isEmpty).orElse(false) && request.limit() == 0) {
+                        Mono.from(gatewayResources.getStore()
+                                .execute(GatewayActions.completeGuildMembers(guildId.asLong())))
+                                .subscribe(null, t -> log.warn("completeGuildMembers action failed for guild {}",
+                                        guildId.asString()));
+                    }
+                });
```

Comment:
I forgot what's the syntax to add behavior on complete while staying in the same chain. Would .handle or .materialize do it or is there an easier way?

Reply:
Ok I discovered a bug so I went for a slightly different approach. Basically if the caller of `GatewayDiscordClient#requestMembers` cancels the subscription before completion (for example via a `.take(n)` operator), it will never execute the store action even though the request members payload was sent. So what I did is that as soon as the request payload is sent, it will add the nonce to a `Set` that is shared with the `GatewayBootstrap#acquireConnection` code where the main dispatch() subscription happens. Tell me if that's good.

Reply:
I need some feedback from @darichey about the `completeGuildMembers` action, whether is worth it for us to track if all members were loaded reliably. This is part of the discussion of reviving "fully ready or fully ready with members" futures present in v2.

If we go forward with this and since it adds more complexity, it might be worth exploring applying this to the `asyncMemberChunk` sequence triggered by `GuildDispatchHandlers`. That is also a limit=0 empty query request.

Also your thoughts on the location of the new `instanceof GuildMembersChunk` code, considering `DispatchStoreLayer` also has a special handling for `instanceof GatewayStateChange`. What would be the criteria for placing code under DispatchStoreLayer vs externally?